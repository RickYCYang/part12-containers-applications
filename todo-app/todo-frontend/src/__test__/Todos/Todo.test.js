import { render, screen, cleanup } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import '@testing-library/jest-dom/extend-expect';
import Todo from '../../Todos/Todo';

const completedTodo = {
  done: true,
  text: 'Learn about container Aris',
  __v: 0,
  _id: '6377b5a0dd5d29951032707e',
};

const notCompletedTodo = {
  done: false,
  text: 'Learn about container Aris',
  __v: 0,
  _id: '6377b5a0dd5d29951032707e',
};

describe('<Todo />', () => {
  const mockDeleteTodo = jest.fn();
  const mockCompleteTodo = jest.fn();
  let container;

  afterEach(() => {
    cleanup();
  });

  test('Render completed Todo', () => {
    container = render(
      <Todo
        todo={completedTodo}
        deleteTodo={mockDeleteTodo}
        completeTodo={mockCompleteTodo}
      />
    ).container;
    const todoText = screen.getByText(`${completedTodo.text}`);
    expect(todoText).toBeInTheDocument();
    const statusText = screen.getByText('This todo is done');
    expect(statusText).toBeInTheDocument();
    const deleteButton = screen.getByText('Delete');
    expect(deleteButton).toBeInTheDocument();
    const setDoneButton = screen.queryByText('Set as done');
    expect(setDoneButton).toBe(null);
    expect(container).toMatchSnapshot();
  });

  test('Render not completed Todo', () => {
    container = render(
      <Todo
        todo={notCompletedTodo}
        deleteTodo={mockDeleteTodo}
        completeTodo={mockCompleteTodo}
      />
    ).container;
    const todoText = screen.getByText(`${completedTodo.text}`);
    expect(todoText).toBeInTheDocument();
    const statusText = screen.getByText('This todo is not done');
    expect(statusText).toBeInTheDocument();
    const deleteButton = screen.getByText('Delete');
    expect(deleteButton).toBeInTheDocument();
    const setDoneButton = screen.getByText('Set as done');
    expect(setDoneButton).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  test('Click delete button of not completed todo', () => {
    container = render(
      <Todo
        todo={notCompletedTodo}
        deleteTodo={mockDeleteTodo}
        completeTodo={mockCompleteTodo}
      />
    ).container;

    const deleteButton = screen.getByText('Delete');
    userEvent.click(deleteButton);
    expect(mockDeleteTodo).toHaveBeenCalled();
    expect(mockDeleteTodo).toHaveBeenCalledWith(notCompletedTodo);
  });

  test('Click set as done button of not completed todo', () => {
    container = render(
      <Todo
        todo={notCompletedTodo}
        deleteTodo={mockDeleteTodo}
        completeTodo={mockCompleteTodo}
      />
    ).container;

    const setDoneButton = screen.getByText('Set as done');
    userEvent.click(setDoneButton);
    expect(mockCompleteTodo).toHaveBeenCalled();
    expect(mockCompleteTodo).toHaveBeenCalledWith(notCompletedTodo);
  });
});
