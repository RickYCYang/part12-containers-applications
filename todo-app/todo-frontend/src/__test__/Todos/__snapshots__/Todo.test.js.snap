// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Todo /> Render completed Todo 1`] = `
<div>
  <div
    style="display: flex; justify-content: space-between; max-width: 70%; margin: auto;"
  >
    <span>
      Learn about container Aris
    </span>
    <span>
      This todo is done
    </span>
    <span>
      <button>
         Delete 
      </button>
    </span>
  </div>
</div>
`;

exports[`<Todo /> Render not completed Todo 1`] = `
<div>
  <div
    style="display: flex; justify-content: space-between; max-width: 70%; margin: auto;"
  >
    <span>
      Learn about container Aris
    </span>
    <span>
      This todo is not done
    </span>
    <span>
      <button>
         Delete 
      </button>
      <button>
         Set as done 
      </button>
    </span>
  </div>
</div>
`;
